scrapy/responsetypes.py

```diff
@@ -102,18 +124,61 @@ def from_body(self, body):
         else:
             return self.from_mimetype('text')
 
+    def _guess_content_type(self, headers=None, url=None, filename=None):
+        if headers and b'Content-Type' in headers:
+            return tuple(headers.getlist(b'Content-Type'))
+
+        if headers and b'Content-Disposition' in headers:
+            filename = headers.get(b'Content-Disposition').split(b';')[-1].split(b'=')[-1].strip(b'"\'').decode()
+        elif url:
+            filename = url
+
+        if filename:
+            mimetype, encoding = self.mimetypes.guess_type(filename)
+            if encoding:
+                return (f"application/{encoding}".encode(),)
+            else:
+                return (mimetype.encode(),)
+
+        return None
+
+    def _guess_response_type(self, mime_type):
+        if not mime_type:
+            return Response
+        if is_html_mime_type(mime_type):
+            return HtmlResponse
+        if is_xml_mime_type(mime_type):
+            return XmlResponse
+        if (
+            mime_type.startswith(b'text/')
+            or is_json_mime_type(mime_type)
+            or is_javascript_mime_type(mime_type)
+        ):
+            return TextResponse
+        return Response
+
     def from_args(self, headers=None, url=None, filename=None, body=None):
         """Guess the most appropriate Response class based on
         the given arguments."""
+        if not body:
+            body = b''
+
+        body = body[:RESOURCE_HEADER_BUFFER_LENGTH]
+        contains_binary_bytes = False
+        for index in range(len(body)):
+            if body[index:index + 1] != b"\x00" and contains_binary(body[index:index + 1]):
+                contains_binary_bytes = True
+                break
+
+        if not contains_binary_bytes:
+            body = body.replace(b"\x00", b"")
```

Comment:
What's the reason for removing them (and not counting them as binary)?

Reply:
For `xtractmime`, a NULL byte in a body causes the body to be interpreted as binary. Please refer [this comment ](https://github.com/scrapy/scrapy/pull/5204#discussion_r676379788)for an example